### DISCUSSION TITLE
onError option in mutate function is not overriding

### DISCUSSION THREAD

**Discussion by QzCurious** [TanStack/query#4599] [Q&A] [answered]
## Mutation function is not overriding `useMutation` and  `defaultOptions`

[![image](https://user-images.githubusercontent.com/38932402/205816200-289354fd-f2c9-431b-895e-83717541715f.png)](https://tanstack.com/query/v4/docs/reference/useMutation)

According to doc, `onError` can override `useMutation`. But this [codesandbox](https://codesandbox.io/s/react-query-onerror-of-mutate-not-overriding-nrounk?file=/src/App.tsx) is not act like that. I'm wondering if it's a bug or the reason of this behavior.

About the [codesandbox](https://codesandbox.io/s/react-query-onerror-of-mutate-not-overriding-nrounk?file=/src/App.tsx):
1. Click the **mutate** button
2. In console, `"onError of useMutation"` is logged first and then `"onError of mutate function"`

## Sidenote

[![image](https://user-images.githubusercontent.com/38932402/205817218-1662f18b-155b-4678-bb4a-61924d8c8a4f.png)](https://tanstack.com/query/v4/docs/reference/QueryCache#global-callbacks)

`onError` of `useMutation` is correctly overriding `defaultOptions`.

**Comment by TkDodo** [comment, marked as answer]
the callbacks on `mutate` are _additional_ to the ones on `useMutation`. They also behave a bit differently:

- https://tanstack.com/query/v4/docs/guides/mutations#mutation-side-effects
- https://tkdodo.eu/blog/mastering-mutations-in-react-query#some-callbacks-might-not-fire

I'd accept a PR to the docs to change this

**Reply by QzCurious** [reply]
Thanks your explanation and information.

I don'k know if there were some misunderstanding for my English. In [Mutation Side Effects](https://tanstack.com/query/v4/docs/guides/mutations#mutation-side-effects), it's also state that `mutate` supports **overrides** on `onSuccess`, `onError` and `onSettled`. Should this be rephrase, too?
![image](https://user-images.githubusercontent.com/38932402/205918279-2a6a7f25-441c-4d7c-a119-423928bb0bae.png)

**Reply by TkDodo** [reply]
`Supported options include` is what it should be :)

**Reply by QzCurious** [reply]
@TkDodo I'm very unfamiliar with custom build project (against to CRA or next.js). I've read though [CONTRIBUTING.md](https://github.com/TanStack/query/blob/9b21609350d9ad41faf5c2fb0ac6c67e4324a637/CONTRIBUTING.md), taking a look what's in [package.json](https://github.com/TanStack/query/blob/9b21609350d9ad41faf5c2fb0ac6c67e4324a637/package.json). But I'm not be able to run a local doc website to start edit it.
(If it's make you easier to make the PR by your self, I can understand it.)

**Reply by TkDodo** [reply]
Just change the markdown file and open a PR :)

**Reply by QzCurious** [reply]
I've created a PR #4601. Hope it helps.

**Comment by QzCurious** [comment]
Doc has been updated by #4599
